### PR TITLE
 Add rule for list item style

### LINKED ISSUE #5: List item style rule
The style is:

```
- [Name](Link) - Description.
```

The rule should:
- Ensure the link and description are delimited with `-`.
- Ensure the description starts with an uppercase.
- Ensure the description ends in a dot `.`.

### PR DESCRIPTION
**Pull request by matheuss**
Closes #5.
- [x] Items in a list must be preceded by a `-`
- [x] A `-` must be present between the link and the description
- [x] The description must start with:
  - [x] an uppercase
  - [x] a camelCase word
  - [x] with a ``code``
- [x] The description must end with `.` or `!`

Also: I'm more than open to suggestions for the error messages 😅

### CODE REVIEW

**Comment by matheuss** [comment]
We have two problems:
1. Seems like `remark-parse`(not sure if I'm blaming the correct module) does not support sublists when the indenting is done with `tabs` :/ – @wooorm any insight on this? Am I doing something wrong? 😰
2. Because of `1.`, this rule will fail when analysing a `Contents` section :(

**Comment by matheuss** [comment]
```js
const visit = require('unist-util-visit');

module.exports = (ast, file, language, done) => {
	visit(ast, 'list', () => {
		console.log('one list here');
	});
	done();
};
```

`tabs.md`:
tabs

`spaces.md`:
spaces

Output:
```
❯ ~ awesome-lint tabs.md  
one list here
❯ ~ awesome-lint spaces.md
one list here
one list here
```

**Comment by matheuss** [comment]
```js
const visit = require('unist-util-visit');

module.exports = (ast, file, language, done) => {
	visit(ast, 'list', list => {
		for (const item of list.children) {
			console.log(JSON.stringify(item.children[1].value));
		}
	});
	done();
};
```

Output: 

```
❯ ~ awesome-lint tabs.md
"\t- [Mad science](#mad-science)\n\t- [Command-line apps](#command-line-apps)"
```

**Comment by wooorm** [comment]
> 1. Seems like remark-parse(not sure if I'm blaming the correct module) does not support sublists when the indenting is done with tabs :/ – @wooorm any insight on this? Am I doing something wrong? 😰

Nope, that’s a bug in remark-parse. 😬 

P.S. It makes it easier (for me) to debug your examples if you add them in code, instead of images, now I have to type them over!

**Comment by matheuss** [comment]
Ops, sorry 😅 Since I needed  show the spaces and tabs, I thought it was better to add them in images — the actual content of the list doesn't matter 😅

File: rules/list-item.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	visit(ast, 'list', list => {
+		let position;
+		for (const item of list.children) {
+			position = JSON.parse(JSON.stringify(item.position));
+			if (position.start.line !== position.end.line) {
+				position.start.line++;
+				file.warn('Items must start with \'- [name][link]\'', position);
+			}
+
+			try { // TODO: seems like unist-util-visit does not supports sublists
```
Comment (sindresorhus) [inline rules/list-item.js]: Can you add a link to the issue https://github.com/wooorm/remark/issues/198 here?

Comment (sindresorhus) [inline rules/list-item.js]: It's not clear from your comment if you work around that issue or not.

File: rules/list-item.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const visit = require('unist-util-visit');
+
+module.exports = (ast, file, language, done) => {
+	visit(ast, 'list', list => {
+		let position;
+		for (const item of list.children) {
+			position = JSON.parse(JSON.stringify(item.position));
+			if (position.start.line !== position.end.line) {
+				position.start.line++;
+				file.warn('Items must start with \'- [name][link]\'', position);
```
Comment (sindresorhus) [inline rules/list-item.js]:
```
file.warn('Items must start with `- [name](link)`', position);
```

File: rules/list-item.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			try { // TODO: seems like unist-util-visit does not supports sublists
+				const description = item.children[0].children.slice(1).reduce((prev, cur) => {
+					if (cur.type === 'inlineCode') {
+						return prev + `\`${cur.value}\``;
+					}
+					return prev + cur.value;
+				}, '');
+
+				if (!description.startsWith(' - ')) {
+					file.warn('Items must have a \'-\' between the link and the description', position);
```
Comment (sindresorhus) [inline rules/list-item.js]:
```
file.warn('List items must have a ` - ` between the link and the description', position);
```

File: rules/list-item.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+				}
+
+				const firstWorld = description.split(' ')[2];
```
Comment (sindresorhus) [inline rules/list-item.js]: World => word

File: rules/list-item.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+				if (['camel', 'capital', 'constant'].indexOf(caseOf(firstWorld)) === -1) {
+					if (!firstWorld.startsWith('`')) {
+						file.warn('The description must start with an uppercase, camelCase world or `code`', position);
```
Comment (sindresorhus) [inline rules/list-item.js]: World => word

File: rules/list-item.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+					}
+				}
+
+				if (!description.endsWith('.') && !description.endsWith('!')) {
+					file.warn('The description must end with a \'.\' or \'!\'', position);
```
Comment (sindresorhus) [inline rules/list-item.js]:
```
file.warn('The description must end with a `.` or `!`', position);
```

File: test/rules.js [hunk 7]
```diff
@@ -12,3 +12,35 @@ test('badge - incorrect source', async t => {
 	t.is(result.ruleId, 'awesome-badge');
 	t.is(result.message, 'Incorrect badge source');
 });
+
+test('list-item - incorrect item prefix', async t => {
+	const results = (await m({filename: 'fixtures/list-item/1.md'})).messages;
+	for (const result of results) {
```
Comment (sindresorhus) [inline test/rules.js]: This is a very fragile way to test it. There might at some point be other violations in the fixtures, so the messages array might also contain other violations. Better to use Array.filter to get the `awesome-list-item` rule and then check the message.

**Comment by sindresorhus** [comment]
For posterity, the referenced Remark issue: https://github.com/wooorm/remark/issues/198

**Comment by matheuss** [comment]
Thanks for the feedback! Already fixed the minor issues ☺️ Now I'm working on the wooorm/remark#198 thing. 
For now, I think that we'll need to `file.contents = file.contents.replace(/\t/g, '  ');` (2 spaces) in `index.js`. Not the best thing in the world, but I think will work for now. 👍 ?

**Comment by sindresorhus** [comment]
Should probably be a bit more specific:

```js
file.contents = file.contents.replace(/^\t- /gm, '  - ');
```

**Comment by matheuss** [comment]
Should we allow this?
```
- [foo](bar) - Baz.
	- [foo](bar) - Nested list item.
	- [foo](bar) - Nested list item.
		- [foo](bar) - Nested list item.
- [foo](bar) - Baz.
```
If yes, the regex you proposed will fail 😰 If no, 👍

**Comment by sindresorhus** [comment]
The regex was just an example. That should indeed be supported. Just adapt the regex for any amount of tab indents.

File: rules/util.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+		const link = node.children[0].children[0];
+		if (link.type === 'link') {
+			return link.url;
+		}
+	} catch (err) {
+		// link not found
+	}
+}
+
+module.exports.getUrlFromItem = getUrlFromItem;
```
Comment (matheuss) [inline rules/util.js]: Quick note: this function is here because it will be useful in the `Table of Contents` rule ( #11 ).

**Comment by matheuss** [comment]
Thanks for the detailed review, Sindre! I think I solved the problems we encountered. The code is a little simpler now and we have more tests ☺️

File: index.js [hunk 9]
```diff
@@ -21,6 +21,11 @@ const m = module.exports = opts => {
 	const run = remark().use(remarkLint, config).process;
 	const file = toVfile.readSync(readmeFile);
 
+	// TODO: temporary workaround for https://github.com/wooorm/remark/issues/198
+	file.contents = file.contents.replace(/^(\t+)- /gm, (_, tabs) => {
+		return `${'  '.repeat(tabs.length)}- `;
+	});
```
Comment (sindresorhus) [inline index.js]:
```js
file.contents = file.contents.replace(/^(\t+)- /gm, (_, tabs) => `${'  '.repeat(tabs.length)}- `);
```

File: test/rules.js [hunk 10]
```diff
@@ -12,3 +12,50 @@ test('badge - incorrect source', async t => {
 	t.is(result.ruleId, 'awesome-badge');
 	t.is(result.message, 'Incorrect badge source');
 });
+
+test('list-item - incorrect item prefix', async t => {
+	t.plan(2);
+	let messages = (await m({filename: 'fixtures/list-item/1.md'})).messages;
+	messages = messages.filter(message => message.ruleId === 'awesome-list-item');
+	messages.forEach(message => t.is(message.message, 'List items must start with `- [name](link)`'));
```
Comment (sindresorhus) [inline test/rules.js]: Use a for-of loop instead

File: test/rules.js [hunk 11]
```diff
@@ -12,3 +12,50 @@ test('badge - incorrect source', async t => {
...
+	t.is(messages.length, 0);
+});
+
+test('list-item – `Contents` section', async t => {
+	let messages = (await m({filename: 'fixtures/list-item/6.md'})).messages;
+	messages = messages.filter(message => message.ruleId === 'awesome-list-item');
+	t.is(messages.length, 0);
+});
+
+test('list-item – item w/o description', async t => {
```
Comment (sindresorhus) [inline test/rules.js]: `w/o` => `without`

File: test/rules.js [hunk 12]
```diff
@@ -12,3 +12,50 @@ test('badge - incorrect source', async t => {
...
+	t.plan(2);
+	let messages = (await m({filename: 'fixtures/list-item/7.md'})).messages;
+	messages = messages.filter(message => message.ruleId === 'awesome-list-item');
```
Comment (sindresorhus) [inline test/rules.js]: These two lines are repeated a lot. Maybe create a helper function.

```js
const messages = run('fixtures/list-item/7.md', 'awesome-list-item');
```

File: test/rules.js [hunk 13]
```diff
@@ -12,3 +12,50 @@ test('badge - incorrect source', async t => {
...
+
+test('list-item - incorrect item prefix', async t => {
+	t.plan(2);
+	let messages = (await m({filename: 'fixtures/list-item/1.md'})).messages;
+	messages = messages.filter(message => message.ruleId === 'awesome-list-item');
+	messages.forEach(message => t.is(message.message, 'List items must start with `- [name](link)`'));
+});
+
+test('list-item - missing dash between link and description', async t => {
+	t.plan(1);
```
Comment (sindresorhus) [inline test/rules.js]: No point in using `t.plan` for sync tests.

File: rules/util.js (hunk 8 above)
Comment (sindresorhus) [inline rules/util.js]: `module.exports.getUrlFromItem` => `exports.getUrlFromItem`

Comment (matheuss) [inline rules/util.js]: ✅

File: test/rules.js (hunk 13 above)
Comment (matheuss) [inline test/rules.js]: I was supposed to ask you this but I forgot: so instead of `t.plan(2)` is better to check the length of the `messages`?

**Comment by matheuss** [comment]
Done ✅

File: rules/list-item.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+			if (position.start.line !== position.end.line) {
+				try {
+					const child = item.children && item.children[1];
+					if (child.type !== 'list') {
+						// TODO: check if the execution somehow reaches this line
+						// probably it will always fall into the catch
+						position.start.line++;
+						file.warn('List items must start with `- [name](link)`', position);
+					}
+				} catch (err) {
```
Comment (sindresorhus) [inline rules/list-item.js]: Why/when would it throw?

File: test/rules.js [hunk 15]
```diff
@@ -1,6 +1,11 @@
 import test from 'ava';
 import m from '../';
 
+async function run(fileName, ruleId) {
+	let results = (await m({filename: fileName})).messages;
```
Comment (sindresorhus) [inline test/rules.js]: Use `const` whenever possible. Applies other places too.

**Comment by sindresorhus** [comment]
@matheuss Can you fix the merge conflict?

**Comment by sindresorhus** [comment]
@wooorm Does it look good to you? Anything that should be done differently/better?

**Comment by sindresorhus** [comment]
```markdown
## foo

- [electron-kr](http://www.meetup.com/electronkr) *(Korean)*
```

Causes:

```
TypeError: Cannot read property 'startsWith' of undefined
    at /Users/sindresorhus/dev/awesome-lint/rules/list-item.js:53:20
    at one (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:72:22)
    at all (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:48:26)
    at one (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:76:20)
    at visit (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:82:5)
    at module.exports.list (/Users/sindresorhus/dev/awesome-lint/rules/list-item.js:8:2)
    at /Users/sindresorhus/dev/awesome-lint/node_modules/wrapped/index.js:74:16
    at wrap (/Users/sindresorhus/dev/awesome-lint/node_modules/wrapped/index.js:54:27)
    at /Users/sindresorhus/dev/awesome-lint/node_modules/remark-lint/lib/index.js:170:5
    at wrapped (/Users/sindresorhus/dev/awesome-lint/node_modules/trough/index.js:128:19)
```

**Comment by sindresorhus** [comment]
```markdown
- [Mancy](https://github.com/princejwesley/Mancy) - JavaScript REPL.
```

Causes a false positive:

```
readme.md
   1:1-1:69  error    The description must start with an uppercase, camelCase word or `code`  awesome-list-item

✖ 1 error
```

**Comment by sindresorhus** [comment]
> file.warn('List items must have a description', position);

It doesn't always make sense to have a description. Sometimes the titles are self-explanatory, like with Articles. Not sure how we should handle that. Maybe drop this warning? Or maybe only show it if other items in a list has it?

**Comment by wooorm** [comment]
I'm not on my computer now, I can do a proper review tomorrow morning though. For now, main concerns are:

* Why is 'position' stringified/parsed?
* what's the deal with try/catch? I like to use something like 'if (node && node.children && node.children[0] && ...)' instead;
* I personally prefer not to use new fns like 'startsWith', especially this case could just use 'charAt(0)', but then this isn't my project!

**Comment by matheuss** [comment]
> Can you fix the merge conflict?

Done 🤗

[Fixed](https://github.com/sindresorhus/awesome-lint/pull/9/commits/d52d03439a4d6016109395d744aa78dc99f8c829) the case when an exception was being thrown and [the one](https://github.com/sindresorhus/awesome-lint/pull/9/commits/0a8c6687c37363e047c65c2a363ffd2ed46fe86c) with the false positive (turns out that `JavaScript` is Pascal case 😛).

About items/lists without description: I was only allowing  a `Contents` section to have items without descriptions. Now [I changed it to allow any list](https://github.com/sindresorhus/awesome-lint/pull/9/commits/d744c1e370708b11f2506cef187e1564f1e07cd8) :) The only requirement is that all items above must have a description.

About the `try-catch`: to be honest, I don't remember. I'll check tomorrow – I certain that I can change it to what @wooorm suggested.

> Why is 'position' stringified/parsed?

I mistook the context 😛 I was afraid of modifying the `position` since other rules may use it. But the modification is inside of a `for (const item of list)`, so I can modify it with no side effects ☺️

**Comment by matheuss** [comment]
Ops, the build failed on Node 4 😢 I'm using [ES2016/17's `includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) which isn't available on Node 4 😕 Do you want me to drop it? Or you would drop Node 4? 😅

**Comment by sindresorhus** [comment]
> I'm using ES2016/17's includes which isn't available on Node 4

Just use `.indexOf`. Not worth dropping Node.js 4 support just for that.

> oh, wait: this is on a test file! Shouldn't AVA take care of that?

AVA did so previously, the problem is that if AVA polyfills it for the tests, it taints all the code running and might change some assumptions. So we turned it off.

**Comment by matheuss** [comment]
Done 😌 

The `try-catch` isn't around anymore: now we have a [safer check ](https://github.com/sindresorhus/awesome-lint/pull/9/commits/25b6c4a60abe50ba96c941ecd56a0e238fd0cee5).

**Comment by sindresorhus** [comment]
```
## Polyfills

- [es6-shim](https://github.com/paulmillr/es6-shim) - Collection of ES2015 polyfills.
- More ES2015 polyfills at [es6-tools](https://github.com/addyosmani/es6-tools#polyfills).
```

Causes:

```
TypeError: Cannot read property 'startsWith' of undefined
    at /Users/sindresorhus/dev/awesome-lint/rules/list-item.js:47:20
    at one (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:72:22)
    at all (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:48:26)
    at one (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:76:20)
    at visit (/Users/sindresorhus/dev/awesome-lint/node_modules/unist-util-visit/index.js:82:5)
    at module.exports.list (/Users/sindresorhus/dev/awesome-lint/rules/list-item.js:7:2)
    at /Users/sindresorhus/dev/awesome-lint/node_modules/wrapped/index.js:74:16
    at wrap (/Users/sindresorhus/dev/awesome-lint/node_modules/wrapped/index.js:54:27)
    at /Users/sindresorhus/dev/awesome-lint/node_modules/remark-lint/lib/index.js:170:5
    at wrapped (/Users/sindresorhus/dev/awesome-lint/node_modules/trough/index.js:128:19)
```

If the list item doesn't start with a link it should just be ignored.

**Comment by matheuss** [comment]
Hm, didn't know that this is allowed. I'll fix it 👌

**Comment by sindresorhus** [comment]
```
- [npms](https://npms.io) - Superb package search with deep analysis of package quality using a [myriad of metrics](https://npms.io/about).
```

Results in:

```
  ✖  809:1   The description of a list item must contain only plain text and/or code  awesome/list-item
```

I think links should be allowed in descriptions.

**Comment by sindresorhus** [comment]
```
- [banshee](https://github.com/eleme/banshee) - Real-time anomalies detection for periodic time series.
```

Results in:

```
✖  783:1   The description must start with an uppercase, camelCase word or code     awesome/list-item
``

I don't see why that is an error.

**Comment by sindresorhus** [comment]
Try running it against https://github.com/sindresorhus/awesome-nodejs

File: rules/list-item.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+					file.warn('List items must start with `- [name](link)`', position);
+				} // else: children === [{type: 'paragraph'...}, {type: 'list'...}] (item with a sublist)
+			}
+			const description = item.children[0].children.slice(1).reduce((prev, cur) => {
+				// we need to `slice(1)` the children so we can discard the link (`[name](link)`)
+				if (cur.type === 'inlineCode') {
+					return `${prev}\`${cur.value}\``;
+				}
+				if (cur.type !== 'text') {
+					file.warn('The description of a list item must contain only plain text and/or `code`', position);
```
Comment (sindresorhus) [inline rules/list-item.js]: remove the backticks from `code`

File: rules/list-item.js [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+				let firstWord = description.split(' ')[2]; // ` - desc`.split(' ') === ['', '-', 'desc']
+				if (firstWord === undefined) {
+					// when link and description are separated with only a ` ` && the description
+					// has only one word
+					firstWord = description.split(' ')[1]; // ` desc`.split(' ') === ['', 'desc']
+				}
+				if (['camel', 'capital', 'constant', 'pascal'].indexOf(caseOf(firstWord)) === -1) {
+					if (!firstWord.startsWith('`')) {
+						file.warn('The description must start with an uppercase, camelCase word or `code`', position);
```
Comment (sindresorhus) [inline rules/list-item.js]: remove the backticks from `code`

**Comment by sindresorhus** [comment]
@matheuss Friendly ping :)

**Comment by matheuss** [comment]
@sindresorhus I was travelling 😌 Tomorrow I'll finish it 💪 😌

**Comment by sindresorhus** [comment]
🍰

**Comment by matheuss** [comment]
I'm sorry @sindresorhus 😔 I'm busy lately, focusing on a project with some friends. 
I'll not promise an ETA, but I'll promise that I'll try to finish it today 😰 😄

**Comment by sindresorhus** [comment]
🦄

**Comment by sindresorhus** [comment]
Closing for lack of activity. Happy to reopen if you ever get to this @matheuss :)

**Pull request closed without merging**